### DISCUSSION TITLE
Protocol extension: read C strings

### DISCUSSION THREAD

**Discussion by pcc** [ARMmbed/DAPLink#1055] [Ideas]
Hi,

I've mentioned this idea elsewhere [1], but I wanted to also post it here to give it some visibility.

A common debugger operation involves reading a C string from memory. With a typical debug adapter protocol this requires a lot of back and forth over USB to avoid reading more data than necessary. In the worst case, we have a round trip per character read. This can generally be optimized to have the debugger software read more than 1 character per round trip, but at the cost of performance for shorter strings.

My idea would be to add an extension to the CMSIS-DAP protocol to support reading null terminated strings of arbitrary length with a single USB round trip. With this extension, the adapter would automatically issue MEM-AP commands to read a null terminated string until reaching the null terminator or a specified maximum length. Then it would send the string over the wire to the host. Host-side debugger software could be taught to use this protocol extension if available.

My questions are:
1. Does this seem like a reasonable extension to CMSIS-DAP?
2. What would be the right place to propose CMSIS-DAP protocol extensions? Would this forum work?

[1] https://github.com/osandov/drgn/pull/313

**Comment by mathias-arm** [comment]
DAPLink is a downstream of [CMSIS-DAP Firmware](https://github.com/ARM-software/CMSIS_5/tree/develop/CMSIS/DAP/Firmware).

You could probably use DAPLink to create a prototype for your command extension(s) by adding it in the [DAPLink vendor command range](https://github.com/ARMmbed/DAPLink/blob/main/source/daplink/cmsis-dap/DAP_vendor.c) or vendor extended command range (see [example here](https://github.com/ARMmbed/DAPLink/blob/develop/source/board/udb/source/DAP_vendor_ex.c#L93) which is board-specific).
